### ISSUE TITLE
[DISCUSSION] Make memory allocation less tied to buffers / textures creation

### ISSUE THREAD

**Issue by dzhdanNV** [NVIDIAGameWorks/NRI#81]
Current usage assumes:
- calling `GetBufferMemoryInfo` and/or `GetTextureMemoryInfo` for buffers and textures
- grouping by abstract `MemoryType`
- calling `AllocateMemory` for a specific `MemoryType`

It doesn't suit well for a case when a chunk of memory needs to be allocated beforehand. How to improve?

Proposed changes:
- NRI: expose:
```c++
enum  MemoryKind {
    ONLY_BUFFERS,
    ONLY_TEXTURES,
    ONLY_COLOR_AND_DEPTH_STENCIL_ATTACHMENTS,
    ALL, // Resource heap Tier2 required
};
```
- NRI: somehow expose resource heap Tier2 support in `DeviceDesc`
- NRI: somehow replace `AllocateMemory` with 2 variants:
   - `AllocateMemoryByType` (current approach)
   - `AllocateMemoryByKind` (new approach: `MemoryKind + MemoryLocation`)
- D3D12: should handle resource heap `Tier1` and `Tier2` under the hood when chooses `MemoryType` and marks for "required dedicated allocation"

**Comment by dzhdanNV** [comment]
@vertver FYI (since you asked for improvements)

**Comment by dzhdanNV** [comment]
VK is problematic:

`vkGet[Buffer/Texture]MemoryRequirements2` is needed to get `memoryRequirements.memoryTypeBits`, this tells which memory types (heaps) can be used for creation. We still can translate `MemoryLocation` to a VK memory type (see `DeviceVK::m_MemoryProps.memoryTypes`), but it will still require a sacral knowledge about resource-heap compatibility. For example, textures can't be created in `DEVICE_UPLOAD` heap (ReBAR).

Additionally, how to check resource heap tier (or something similar) in VK?

**Comment by dzhdanNV** [comment]
Good comparison of VK and D3D12 in the light we need. At least 2 things to note:
- probably it's time to stop using `vkGet[Buffer/Image]MemoryRequirements2` (requiring a resource handle), and replace them with `vkGetDevice[Buffer/Image]MemoryRequirements` expecting a resource description instead (like in D3D12!)
- it looks like `MemoryKind` can be just ignored in VK

**Comment by dzhdanNV** [comment]
Currently working on changing:
```c++
    void (NRI_CALL *GetBufferMemoryInfo)(const NRI_NAME_REF(Buffer) buffer, NRI_NAME(MemoryLocation) memoryLocation, NRI_NAME_REF(MemoryDesc) memoryDesc);
    void (NRI_CALL *GetTextureMemoryInfo)(const NRI_NAME_REF(Texture) texture, NRI_NAME(MemoryLocation) memoryLocation, NRI_NAME_REF(MemoryDesc) memoryDesc);
```
to:
```c++
    void (NRI_CALL *GetBufferMemoryDesc)(const NRI_NAME_REF(Device) device, const NRI_NAME_REF(BufferDesc) bufferDesc, NRI_NAME(MemoryLocation) memoryLocation, NRI_NAME_REF(MemoryDesc) memoryDesc);
    void (NRI_CALL *GetTextureMemoryDesc)(const NRI_NAME_REF(Device) device, const NRI_NAME_REF(TextureDesc) textureDesc, NRI_NAME(MemoryLocation) memoryLocation, NRI_NAME_REF(MemoryDesc) memoryDesc);
```
It unlocks a ton of flexibility, eliminated the need to "create" a buffer / texture object first. Also, despite it's a breaking change, the old code still can be used with "no brain" modifications, because resource `desc` can be retrieved from an already created `Buffer` or `Texture` object at any time.

VK implementation has switched to modern functions from `VK_KHR_maintanence4`:
- `GetBufferMemoryRequirements2` => `GetDeviceBufferMemoryRequirements`
- `GetImageMemoryRequirements2` => `GetDeviceImageMemoryRequirements`

**Comment by vertver** [comment]
Sounds good to me, definitely worth it! Since you don't have to rewrite code, it makes sense to change the API in this way.

**Comment by dzhdanNV** [comment]
Actually this change already makes NRI memory API friendly to render graph based systems, i.e. when some big chunks of memory need to be allocated in advance. But currently it works in this way:
- prerequisite: sacral knowledge about D3D12 resource heap tiers
- call `GetBufferMemoryDesc`
- call `GetTextureMemoryDesc` for a color or depth-stencil attachment texture
- call `GetTextureMemoryDesc` for a shader storage or read-only texture
- check returned `MemoryType`s
  - if they are the same, `tier2` is supported, a single huge heap can be used for everything
  - if they are different, `tier2` is not supported, 3 heaps are needed

Probably, it makes sense to expose something like `DeviceDesc::memoryTier` and add:
```c++
enum MemoryTier {
    MEMORY_TIER1, // at least 3 memory chunks are needed for: buffers, color and depth-stencil attachments and other textures
    MEMORY_TIER2 // at least 1 memory chunk is needed, which can include buffers and textures with any usage flags
};
```
P.S. Looks like VK is always `MEMORY_TIER2` but with a potential need in a "dedicated memory allocation".

**Comment by dzhdanNV** [comment]
Implemented in `v1.138`. NRI samples updated too.
